### PR TITLE
json: make the json streamer a template class

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: json: split low level methods out from json streamer
Additional Description:

This PR split low level methods of JSON serialization out from `Streamer` and make it a template to accept different types of output (Buffer::Instance, std::string, etc.)

Risk Level: low.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

File: source/common/json/json_streamer.h [hunk 1]
```diff
@@ -1,25 +1,119 @@
...
+  // Constructor with output buffer to write the JSON pieces.
+  Serializer(OutputBuffer& output_buffer) : output_buffer_(output_buffer) {}
+
+  /**
+   * Add raw JSON piece to the output buffer. The value could be delimiters,
+   * or sanitized JSON piece.
+   *
+   * @param value The raw JSON piece to be added.
+   */
+  void addPiece(absl::string_view value) { output_buffer_.addFragments({value}); }
```
Comment (wbpcode) [inline source/common/json/json_streamer.h:53]: Like the original `addConstantString`. But it possible for users to add raw JSON piece to the buffer by this method, not just `constant` string. So, I changed the name.

File: source/common/json/json_streamer.h [hunk 2]
```diff
@@ -1,25 +1,119 @@
...
+   * Default be `"`.
+   *
+   * NOTE: Both key and string values should use this method to sanitize.
+   */
+  void addString(absl::string_view value, absl::string_view prefix = Constants::Quote,
+                 absl::string_view suffix = Constants::Quote) {
+    // Sanitize the string value and quote it on demand of the caller.
+    absl::string_view sanitized = Json::sanitize(sanitize_buffer_, value);
+    output_buffer_.addFragments({prefix, sanitized, suffix});
+  }
```
Comment (wbpcode) [inline source/common/json/json_streamer.h:72]: Like previous `addSanitized`. But I changed the order of the input parameter to provide default prefix and suffix (`"`).

File: source/common/json/json_streamer.h [hunk 3]
```diff
@@ -1,25 +1,119 @@
...
+   * @param value The bool value to be added.
+   */
+  void addBool(bool value) {
+    output_buffer_.addFragments({value ? Constants::True : Constants::False});
+  }
+
+  /**
+   * Add a null value to the raw JSON piece buffer.
+   */
+  void addNull() { output_buffer_.addFragments({Constants::Null}); }
```
Comment (wbpcode) [inline source/common/json/json_streamer.h:102]: These are completely same with original impl except the `output_buffer_` is templated.

File: source/common/json/json_sanitizer.cc [hunk 4]
```diff
@@ -65,7 +65,7 @@ absl::string_view sanitize(std::string& buffer, absl::string_view str) {
   if (need_slow == 0) {
     return str; // Fast path, should be executed most of the time.
   }
-  TRY_ASSERT_MAIN_THREAD {
+  try {
```
Comment (jmarantz) [inline source/common/json/json_sanitizer.cc:68]: why this change? It will fail lint (as it should).

File: source/common/json/json_streamer.cc [hunk 5]
```diff
@@ -1,199 +1,3 @@
...
-void Streamer::addBool(bool b) { response_.addFragments({b ? "true" : "false"}); }
-
-void Streamer::addSanitized(absl::string_view prefix, absl::string_view str,
-                            absl::string_view suffix) {
-  absl::string_view sanitized = Json::sanitize(sanitize_buffer_, str);
-  response_.addFragments({prefix, sanitized, suffix});
-}
-
-} // namespace Json
+namespace Json {} // namespace Json
```
Comment (jmarantz) [inline source/common/json/json_streamer.cc:2]: the way you structured this PR Is maybe a better end result but is hard to review.

Can you follow how I did this in #35647 -- that shows what functionality actually needed to change.

A follow-up could be done after that which is completely non-functional to move the contents of the cc file into the .h.

File: source/common/json/json_sanitizer.cc (hunk 4 above)
Comment (wbpcode) [reply source/common/json/json_sanitizer.cc:68]: we will construct json log line at worker threads, so this change is necessary.

File: source/common/json/json_streamer.cc (hunk 5 above)
Comment (wbpcode) [reply source/common/json/json_streamer.cc:2]: Okay

Comment (wbpcode) [reply source/common/json/json_streamer.cc:2]: Okay

File: source/common/json/json_sanitizer.cc (hunk 4 above)
Comment (jmarantz) [reply source/common/json/json_sanitizer.cc:68]: I think a separate change is needed to solve this problem. This won't pass lint; we can't catch exceptions in worker threads.

File: source/common/json/json_streamer.h (hunk 1 above)
Comment (jmarantz) [reply source/common/json/json_streamer.h:88]: suggest calling it addWIthoutSanitizing or something like that to reflect that you are bypassing the sanitizer -- that's the main reason to use it, right?

Comment (jmarantz) [reply source/common/json/json_streamer.h:88]: or addStringWithoutSanitizing.

File: tools/code_format/config.yaml [hunk 6]
```diff
@@ -231,6 +231,7 @@ paths:
     include:
     - source/common/common/thread.h
     - source/exe/terminate_handler.cc
+    - source/common/json/json_sanitizer.cc
```
Comment (jmarantz) [inline tools/code_format/config.yaml:234]: let's not do this; this is really a violation and we shouldn't just ignore it.

File: source/common/json/json_sanitizer.cc (hunk 4 above)
Comment (wbpcode) [reply source/common/json/json_sanitizer.cc:68]: How? Make this exception free directly? My initial plan was to create an exception free version sanitize method. But you hope to defer that to future and take much time to convince me.

File: tools/code_format/config.yaml (hunk 6 above)
Comment (wbpcode) [reply tools/code_format/config.yaml:234]: I think we back to the previous problem: how to handle the exception.

File: source/common/json/json_streamer.h (hunk 1 above)
Comment (wbpcode) [reply source/common/json/json_streamer.h]: Hah, I just see this comment. Because when we adding a string, we will surround it with quote. So, the method is not just without sanitizing, but do nothing additionally.
So, how about to name it `addDirectly` means that will put a string to buffer directly  and no any other operations.

**Comment by wbpcode** [comment]
I still think it's weird to include .cc in a .h file (Although I know there are some similar cases in real world. But typically we can avoid that here).

To simplify the review, I split the PR to clean 3 commits:
1. make the Streamer a template class like https://github.com/envoyproxy/envoy/pull/35647
2. remove unnecessary methods and simplify the code.
3. no any logic but move template code from .cc to .h.

You can review them one by one to simplify the review processing. cc @jmarantz

**Comment by wbpcode** [comment]
Force update is used because re-organization of commits.

**Comment by wbpcode** [comment]
The benchmark result of previous stats handler:

```
----------------------------------------------------------------------------------------------
Benchmark                                                    Time             CPU   Iterations
----------------------------------------------------------------------------------------------
BM_AllCountersJson/per_endpoint_stats_disabled             590 ms          590 ms            1 output per iteration: 135789011
[2024-08-28 10:20:15.902][265134][error] [test/server/admin/stats_handler_speed_test.cc:163] Initializing cluster info; slow to construct and destruct...
BM_AllCountersJson/per_endpoint_stats_enabled             1019 ms         1019 ms            1 output per iteration: 180082411
BM_UsedCountersJson/per_endpoint_stats_disabled           14.0 ms         14.0 ms           50 output per iteration: 1348901
BM_UsedCountersJson/per_endpoint_stats_enabled             424 ms          424 ms            2 output per iteration: 45642301
BM_FilteredCountersJson/per_endpoint_stats_disabled        161 ms          161 ms            4 output per iteration: 12
BM_FilteredCountersJson/per_endpoint_stats_enabled         504 ms          504 ms            1 output per iteration: 12
BM_HistogramsJson/per_endpoint_stats_disabled             10.9 ms         10.9 ms           65 output per iteration: 1820890
BM_HistogramsJson/per_endpoint_stats_enabled              10.6 ms         10.6 ms           66 output per iteration: 1820890
```

The benchmark result of stats handler on the latest commit of this PR:

```
----------------------------------------------------------------------------------------------
Benchmark                                                    Time             CPU   Iterations
----------------------------------------------------------------------------------------------
BM_AllCountersJson/per_endpoint_stats_disabled             576 ms          576 ms            1 output per iteration: 135789011
[2024-08-28 14:01:43.273][593717][error] [test/server/admin/stats_handler_speed_test.cc:163] Initializing cluster info; slow to construct and destruct...
BM_AllCountersJson/per_endpoint_stats_enabled              938 ms          938 ms            1 output per iteration: 180082411
BM_UsedCountersJson/per_endpoint_stats_disabled           13.0 ms         13.0 ms           50 output per iteration: 1348901
BM_UsedCountersJson/per_endpoint_stats_enabled             393 ms          393 ms            2 output per iteration: 45642301
BM_FilteredCountersJson/per_endpoint_stats_disabled        161 ms          161 ms            4 output per iteration: 12
BM_FilteredCountersJson/per_endpoint_stats_enabled         457 ms          457 ms            2 output per iteration: 12
BM_HistogramsJson/per_endpoint_stats_disabled             9.65 ms         9.65 ms           69 output per iteration: 1820890
BM_HistogramsJson/per_endpoint_stats_enabled              9.59 ms         9.59 ms           66 output per iteration: 1820890
```

The latest commit provide similar (or a little better) performance.

**Comment by phlax** [comment]
needs main merge - but figure its waiting on other PRs etc

/wait

**Comment by jmarantz** [comment]
I'm anticipating this will now be straightforward once the inlining one is merged.

I had another thought: call the template JsonStreamerTemplate or JsonStreamerBase, and leave JsonStreamer as a nickname for JsonStreamerTemplate<Buffer> or equivalent. That will avoid having to change all the call-sites of the existing JsonStreamer.

**Comment by wbpcode** [comment]
Replaced by #36001
